src/service/viewport-impl.js

```diff
@@ -1093,22 +1137,37 @@ export class ViewportBindingNaturalIosEmbed_ {
   }
 
   /** @override */
-  updatePaddingTop(paddingTop, opt_updateScrollPos, opt_lastPaddingTop) {
+  hideViewerHeader(transient, lastPaddingTop) {
+    if (!transient) {
+      this.updatePaddingTop(0);
+    } else {
+      // Add extra paddingTop to make the content stay at the same position
+      // when the hiding header operation is transient
+      onDocumentReady(this.win.document, doc => {
+        const existingPaddingTop =
+            this.win./*OK*/getComputedStyle(doc.body)['padding-top'] || '0';
+        doc.body.style.paddingTop =
+            `calc(${existingPaddingTop} + ${lastPaddingTop}px)`;
+        doc.body.style.borderTop = '';
+      });
+    }
+  }
+
+  /** @override */
+  showViewerHeader(transient, paddingTop) {
+    if (!transient) {
+      this.updatePaddingTop(paddingTop);
+    }
+    // No need to adjust borderTop and paddingTop when the showing header
+    // operation is transient
+  }
+
+  /** @override */
+  updatePaddingTop(paddingTop) {
     onDocumentReady(this.win.document, doc => {
       this.paddingTop_ = paddingTop;
-      // Also tried `paddingTop` but it didn't work for `position:absolute`
-      // on iOS.
       doc.body.style.borderTop = `${paddingTop}px solid transparent`;
-      if (opt_updateScrollPos) {
-        // TODO(yuxichen): This is a partially working formula for calculating
-        // adjusted scroll top. Add two more paddingTop to compensate the
-        // paddingTop being removed by the setScrollTop function and the removal
-        // of border top. This formula only works when either paddingTop or
-        // lastPaddingTop is 0.
-        const adjScrollTop = this.getScrollTop() + 3 * paddingTop
-            - opt_lastPaddingTop;
-        this.setScrollTop(adjScrollTop);
-      }
+      doc.body.style.paddingTop = '';
```

Comment:
I am calling `updatePaddingTop` when we want to have a space on the top for the header. We use `body.style.borderTop` for the space. And we need to reset the `body.style.paddingTop` to the user-defined padding top by setting it to empty.